EIPS/eip-1283.md

```diff
@@ -0,0 +1,109 @@
+---
+eip: 1283
+title: Net gas metering for SSTORE without dirty maps
+author: Wei Tang (@sorpaas)
+discussions-to: https://github.com/sorpaas/EIPs/issues/1
+status: Draft
+type: Standards Track
+category: Core
+created: 2018-08-01
+---
+
+## Abstract
+
+This EIP proposes net gas metering changes for SSTORE opcode, as an
+alternative for EIP-1087. It tries to be friendlier to implementations
+that uses different opetimiazation strategies for storage change
+caches.
+
+## Motivation
+
+EIP-1087 proposes a way to adjust gas metering for SSTORE opcode,
+enabling new usages on this opcodes where it is previously too
+expensive. However, EIP-1087 requires keeping a dirty map for storage
+changes, and implictly makes the assumption that a transaction's
+storage changes are committed to the storage trie at the end of a
+transaction. This works well for some implementations, but not for
+others. Some implementations do the optimization to only commit
+storage changes at the end of a block. For them, it is possible to
+know a storage's original value and current value, but it is not
+possible to iterate over all storage changes. For EIP-1087, they will
+need to keep a separate dirty map to keep track of gas costs. This
+adds additional memory consumptions.
+
+This EIP proposes an alternative way for gas metering on SSTORE, using
+information that is more universially available to most
+implementations:
+
+* *Storage slot's original value*. This is the value of the storage if
+  a call/create reversion happens on the current VM execution
+  context. It is universially available because all clients need to
+  keep track of call/create reversion.
+* *Storage slot's current value*. 
+* Refund counter.
+
+## Specification
+
+Term *original value* is as defined in Motivation. *Current value*
+refers to the storage slot value before SSTORE happens. *New value*
+refers to the storage slot value after SSTORE happens.
+
+Replace SSTORE opcode gas cost calculation (including refunds) with
+the following logic:
+
+* If *current value* equals *new value* (this is a no-op), 200 gas is
```

Comment:
Following this process, starting with *original* = 1 and *current* = 1:

1)  *new* = 0; deducts 5k gas and adds 15k gas to the refund counter; *current* = 0.
2) *new* = 1; adds 4800 gas to the refund counter
3) Goto 1.

Thus for every 5k gas deducted from the gas meter, we can add 19800 gas to the refund counter.

Reply:
This seems reasonable. A couple of critiques:

 - Unlike the existing proposal, this can result in consuming a lot of gas and adding it to the refund counter. Since refunds are limited to 1/2 the total gas used, you could easily end up paying more than expected if you set a value to and from its original value a number of times.
 - It's hard to follow the logic here. Some diagramming, or breaking down the cases more would help.

Reply:
Thanks! I think (1) is a really good point. If there're cases where it can cost a lot most gases, then we may break backward compatibility for existing contracts. So in https://github.com/ethereum/EIPs/pull/1283/commits/23ebc94bcf5896376e497936a471b63e567bc84f I changed the clause from deducting 15k gas to removing 15k gas from refund counter. In this way, we will never consume more gases compared with current scheme. Indeed, this adds a new semantics for gas refunds (we never removed gas from refund counter before), so I'm definitely open to better ideas on this. The current rationale/assumption for this is:

* Backward compatibility is important.
* We can prove that refund counter will never go below 0.
* This new semantic of removing gas from refund counter is trivial to implement in most clients.

Still working on (2). Let me try to see whether there are clearer ways to describe the spec.

Reply:
Added an explanation section to make the logic more clearer!